python/graphstorm/model/edge_decoder.py

```diff
@@ -713,6 +714,45 @@ def out_dims(self):
         """
         return 1
 
+class LinkPredictContrastiveDotDecoder(LinkPredictDotDecoder):
+    """ Link prediction decoder designed for contrastive loss
+        with the score function of dot product
+    """
+
+    # pylint: disable=unused-argument
+    def forward(self, g, h, e_h=None):
+        with g.local_scope():
+            scores = {}
+
+            for canonical_etype in g.canonical_etypes:
+                if g.num_edges(canonical_etype) == 0:
+                    continue # the block might contain empty edge types
+
+                src_type, _, dest_type = canonical_etype
+                u, v = g.edges(etype=canonical_etype)
+                # Sort edges according to source node ids
+                # The same function is invoked by computing both pos scores
+                # and neg scores, by sorting edges according to source nids
+                # the output scores of pos_score and neg_score are compatible.
+                #
+                # For example:
+                #
+                # pos pairs   |  neg pairs
+                # (10, 20)    |  (10, 3), (10, 1), (10, 0), (10, 22)
+                # (13, 6)     |  (13, 3), (13, 1), (13, 0), (13, 22)
+                # (29, 8)     |  (29, 3), (29, 1), (29, 0), (29, 22)
+                # TODO: use stable to keep the order of negatives. This may not
+                # be necessary
+                u_sort_idx = th.argsort(u, stable=True)
```

Comment:
this is really a temporary solution. the comment above actually doesn't make too much sense to be here because there are no positive and negative pairs in this function. the comment should be in the GNN module where we can see positive edges and negative edges.

Reply:
You are right.
It works only because we know how neg_graph is built.

Reply:
can you add a TODO here?

Reply:
I added a TODO in L724: 
```
TODO(xiang): develop a better solution to unify the implementation of link prediction decoders.
```

Reply:
I added a TODO in L725 (class description).